src/io/file_io.cpp

```diff
@@ -0,0 +1,178 @@
+#include <LightGBM/utils/log.h>
+#include <LightGBM/utils/file_io.h>
+
+#include <algorithm>
+#include <sstream>
+#include <unordered_map>
+#ifdef USE_HDFS
+#include <hdfs.h>
+#endif
+
+namespace LightGBM{
+
+struct LocalFile : VirtualFileReader, VirtualFileWriter {
+  LocalFile(const std::string& filename, const std::string& mode) : filename_(filename), mode_(mode) {}
+  virtual ~LocalFile() {
+    if (file_ != NULL) {
+      fclose(file_);
```

Comment:
set file to NULL after close ? I am not sure will `fclose` change FILE to NULL or not.